folium/folium.py

```diff
@@ -538,55 +556,73 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
-        elif color_data:
-            # To avoid explicit pandas dependency ; changed default behavior.
-            data_min = min(color_data.values())
-            data_max = max(color_data.values())
-            if data_min == data_max:
-                data_min = (data_min if data_min < 0 else 0
-                            if data_min > 0 else -1)
-                data_max = (data_max if data_max > 0 else 0
-                            if data_max < 0 else 1)
-            data_min, data_max = (1.01*data_min-0.01*data_max,
-                                  1.01*data_max-0.01*data_min)
-            nb_class = 6
-            color_domain = [data_min+i*(data_max-data_min)*1./nb_class
-                            for i in range(1+nb_class)]
-        else:
-            color_domain = None
+        if color_data is not None and key_on is not None:
+            # numpy histogram_bin_edges is only available in the most
+            # recent numpy version, so we will use numpy histogram
+            real_values = np.array(list(color_data.values()))
+            real_values = real_values[~np.isnan(real_values)]
+            _, bins = np.histogram(real_values, bins=bins)
+
+            bins_min, bins_max = min(bins), max(bins)
+            if np.any((real_values < bins_min) | (real_values > bins_max)):
+                raise ValueError('''
+                All values are expected to fall into one of the provided bins
+                (or to be Nan). Please check the `bins` parameter
+                and/or your data.''')
+
+            # then we 'correct' the last edge for numpy digitize
+            # (we add a very small amount to fake an inclusive right interval)
+            increasing = bins[0] <= bins[-1]
+            bins[-1] = np.nextafter(
+                bins[-1],
+                (1 if increasing else -1) * np.inf)
 
-        if color_domain and key_on is not None:
             key_on = key_on[8:] if key_on.startswith('feature.') else key_on
-            color_range = color_brewer(fill_color, n=len(color_domain))
+
+            nb_bins = len(bins) - 1
+            color_range = color_brewer(fill_color, n=nb_bins)
 
             def get_by_key(obj, key):
                 return (obj.get(key, None) if len(key.split('.')) <= 1 else
                         get_by_key(obj.get(key.split('.')[0], None),
                                    '.'.join(key.split('.')[1:])))
 
             def color_scale_fun(x):
-                idx = len(
-                    [
-                        u for u in color_domain if
-                        get_by_key(x, key_on) in color_data and
-                        u <= color_data[get_by_key(x, key_on)]
-                    ]
-                )
-                return color_range[idx-1]
+                key_of_x = get_by_key(x, key_on)
+
+                if key_of_x not in color_data.keys():
+                    # 1: opacity is handled within nan_fill_color
+                    return nan_fill_color, 1
```

Comment:
Having `nan_fill_color` as RGBA values adds some complexity. What if someone uses a regular RGB value and expects the `fill_opacity` to work on these fields as well? Do we have to check if `nan_fill_color` is a RGBA value and only return `opacity = 1` in that case?

I'm skeptical of adding more complexity by parsing color strings. Either return `opacity 1` or go back to the RGB situation we had before. What do you think?

Reply:
Or maybe instead add a `nan_fill_opacity` argument? That would be more in line with the rest of the code. We don't use RGBA in any other place I think.

Reply:
I think, the easiest way to go would be to add a `nan_fill_opacity` parameter (that would default to the `fill_opacity` value if it's not set) to the function because I am sure at some point in the future someone is going to ask for such possibility.
And going that way we won't have to think about tricky HEX RGBA strings.

What do you think ?

Reply:
Oh we thought at the same thing... I'll add the parameter.

Reply:
I agree, that would be best. 

I do like the idea of using RGBA though, so thanks for introducing that. But we would have to handle that more consistently then. Maybe for in another PR.

Reply:
I have updated the code to have a `nan_fill_opacity` parameter.